File: 1028-interval-list-intersections/interval-list-intersections.cpp
class Solution {
public:
    vector<vector<int>> intervalIntersection(vector<vector<int>>& firstList, vector<vector<int>>& secondList) {
        vector<vector<int>> result;
         int i = 0; // Pointer for firstList
        int j = 0; // Pointer for secondList

        while (i < firstList.size() && j < secondList.size()) {
            
            // Formula: Max of starts, Min of ends
            int start_time = max(firstList[i][0], secondList[j][0]);
            int end_time = min(firstList[i][1], secondList[j][1]);

            // Agar start_time chhota ya barabar hai end_time ke
            if (start_time <= end_time) {
                result.push_back({start_time, end_time});
            }

            // Pointer increment karne ka dimaag
            if (firstList[i][1] < secondList[j][1]) {
                i++;
            } else {
                j++;
            }
        }

        return result;
    }
};

File: 1028-interval-list-intersections/interval-list-intersections.java
class Solution {
    public int[][] intervalIntersection(int[][] firstList, int[][] secondList) {
         List<int[]> result = new ArrayList<>();
        int i = 0; // Pointer for firstList
        int j = 0; // Pointer for secondList

        // Jab tak dono arrays me elements bache hain
        while (i < firstList.length && j < secondList.length) {
            
            // Formula: Starts ka maximum aur Ends ka minimum
            int start_time = Math.max(firstList[i][0], secondList[j][0]);
            int end_time = Math.min(firstList[i][1], secondList[j][1]);

            // Agar valid common slot hai toh list me daal do
            if (start_time <= end_time) {
                result.add(new int[]{start_time, end_time});
            }

            // Galti yahan hoti hai: Jiska END time chhota hai, use aage badhao
            if (firstList[i][1] < secondList[j][1]) {
                i++; // firstList ka interval pehle khatam hua
            } else {
                j++; // secondList ka interval pehle khatam hua ya barabar hai
            }
        }

        // ArrayList ko 2D array me badal kar return karo
        return result.toArray(new int[result.size()][]);
    }
}